src/plasma_manager.c

```diff
@@ -23,63 +23,262 @@
 
 #include "uthash.h"
 #include "utlist.h"
+#include "utarray.h"
 #include "utstring.h"
 #include "common.h"
 #include "io.h"
 #include "event_loop.h"
 #include "plasma.h"
 #include "plasma_client.h"
 #include "plasma_manager.h"
+#include "state/db.h"
+#include "state/object_table.h"
+
+#define NUM_RETRIES 5
+
+/* Timeouts are in milliseconds. */
+#ifndef RAY_TIMEOUT
+#define MANAGER_TIMEOUT 1000
+#else
+#define MANAGER_TIMEOUT RAY_TIMEOUT
+#endif
+
+typedef struct client_object_connection client_object_connection;
 
 typedef struct {
+  /** Event loop. */
+  event_loop *loop;
   /** Connection to the local plasma store for reading or writing data. */
-  plasma_store_conn *store_conn;
-  /** Hash table of all contexts for active connections to other plasma
-   * managers. These are used for writing data to other plasma stores. */
+  plasma_connection *plasma_conn;
+  /** Hash table of all contexts for active connections to
+   *  other plasma managers. These are used for writing data to
+   *  other plasma stores. */
   client_connection *manager_connections;
+  db_handle *db;
+  /** Our address. */
+  uint8_t addr[4];
+  /** Our port. */
+  int port;
+  /** Hash table of outstanding fetch requests. The key is
+   *  object id, value is a list of connections to the clients
+   *  who are blocking on a fetch of this object. */
+  client_object_connection *fetch_connections;
 } plasma_manager_state;
 
-typedef struct plasma_buffer plasma_buffer;
+plasma_manager_state *g_manager_state = NULL;
+
+typedef struct plasma_request_buffer plasma_request_buffer;
 
-/* Buffer for reading and writing data between plasma managers. */
-struct plasma_buffer {
+/* Buffer for requests between plasma managers. */
+struct plasma_request_buffer {
+  int type;
   object_id object_id;
   uint8_t *data;
   int64_t data_size;
   uint8_t *metadata;
   int64_t metadata_size;
-  int writable;
   /* Pointer to the next buffer that we will write to this plasma manager. This
-   * field is only used if we're transferring data to another plasma manager,
+   * field is only used if we're pushing requests to another plasma manager,
    * not if we are receiving data. */
-  plasma_buffer *next;
+  plasma_request_buffer *next;
+};
+
+/* The context for fetch and wait requests. These are per client, per object. */
```

Comment:
No, there is a new context per object. A request can include multiple objects so I'd prefer not to call it that.